common-npm-packages/azure-arm-rest-v2/azure-arm-common.ts

```diff
@@ -154,12 +154,16 @@ export class ApplicationTokenCredentials {
         return this.clientId;
     }
 
+    public getUseMSAL(): boolean {
```

Comment:
I think that we should not expose this value from credentials. As there is a need to use this value somewhere else (`ServicePrincipals`, `GraphManagementClient`) I suggest to introduce new entity that would provide such information. Such entity might be injected to all places where we need to determine if MSAL is being used. With current setup the code forces us to use credentials whenewer we need to know about MSAL even in case that we do not need anything else from credentials. I believe this breaks single responsibility principle and exposes implementation details.

Reply:
I agree with refactoring. I have a future plan to implement factory design pattern to make it cleaner and get rid of unnecessary complexity. I think it might be a good fit for building these configurations and it would help to get rid of current many constructor parameters.

At this time period; regarding to exposing, I think `getUseMSAL()` would be also valuable for the Tasks to determine whether if MSAL is applied or not as a runtime configuration. `ApplicationTokenCredentials ` itself is an entity which is responsible of token generation but of course it is possible to separate it to smaller responsibilities.